### PR TITLE
Add charter for WG Structured Logging

### PR DESCRIPTION
**Pull request by serathius**
/cc @dims @shubheksha @thockin 
@kubernetes/steering-committee

### CODE REVIEW

File: wg-structured-logging/charter.md
Comment (gkarthiks) [inline wg-structured-logging/charter.md:38]: we are not focusing on CLI?

**Review by portante — COMMENTED** [review]
I'd be happy to participate in this structured logging effort.  Thanks!

**Comment by parispittman** [comment]
lgtm from steering

**Comment by dims** [comment]
LGTM, reads well and is comprehensive to cover the problem areas.

File: wg-structured-logging/charter.md
Comment (portante) [inline wg-structured-logging/charter.md:24]: Could we consider amending this charter to provide tools that out of scope projects could use to apply the last four points above to themselves?

File: wg-structured-logging/charter.md
Comment (portante) [reply wg-structured-logging/charter.md:38]: Seems like `kubectl logs` could be impacted?

File: wg-structured-logging/charter.md
Comment (portante) [inline wg-structured-logging/charter.md:57]: Perhaps, `s/sig follows adheres/sig adheres/`?

File: wg-structured-logging/charter.md
Comment (portante) [inline wg-structured-logging/charter.md:64]: Perhaps, `s/existing/existence/`?

File: wg-structured-logging/charter.md
Comment (portante) [inline wg-structured-logging/charter.md:68]: Perhaps, `s/we will be/is`, and `s/shape of them and no/shape of them, and no`?

File: wg-structured-logging/charter.md
Comment (portante) [inline wg-structured-logging/charter.md:69]: Perhaps, `s/will be/is`, and `s/we will retire Working Group/the Working Group will retire/`?

**Review by portante — CHANGES_REQUESTED** [review]

File: wg-structured-logging/charter.md
Comment (gkarthiks) [reply wg-structured-logging/charter.md:38]: Yeah, that was my initial thought on this. But I'll defer to the people who know much about this.

Comment (serathius) [reply wg-structured-logging/charter.md:38]: Main target will be logs generated by K8s system components (Apiserver, Scheduler, Controller, Kubelet), controllers using `k8s.io/component-base` repository and common libraries used by them. 

Binaries like kubectl will not be directly targeted nor we plan to change application logs that `kubectl logs` reads.

**Comment by serathius** [comment]
@portante Thanks for feedback. Updated the PR.

File: wg-structured-logging/charter.md
Comment (liggitt) [inline wg-structured-logging/charter.md:28]: where does logging done by upstream dependencies fit? is everything that contributes to the logfile of core Kubernetes components in-scope if the entire log file is expected to be structured?

File: wg-structured-logging/charter.md
Comment (liggitt) [inline wg-structured-logging/charter.md:50]: this was explicitly a non-goal of the structured logging effort ("We are not replacing currently used logging library (klog) or the way in which it is used") and the `glog → klog` and `klog → klog/v2` were significant efforts that rippled across many of our upstream dependencies... does including this in the charter commit the project to this as a goal or is this still under discussion?

File: wg-structured-logging/charter.md
Comment (serathius) [reply wg-structured-logging/charter.md:28]: Yes, I think that covering upstream dependencies is needed to have logs in single format which is required to achieve goal of making logs easy to consume. 

To make sure I understood you correctly, when talking about "upstream" dependencies do you mean:
* Kubernetes shared libraries in ./staging/src/k8s.io directory? or
* Third party libraries outside of k/k repo?

File: wg-structured-logging/charter.md
Comment (serathius) [reply wg-structured-logging/charter.md:50]: Original structured logging proposal was much more ambitious with goals outside just interface change. With creation of working group I think it's important to revisit those ideas and finally solve those long hanging problems introduced originally by glog and are still present in klog (global confusing flags, global state). Based on my discussion with @thockin, klog is not very favored dependency In golang community. Even though replacing klog will require a lot of cost from whole community I think it's still we welcomed as an big improvement

/cc @thockin 
Who has more strong opinion about forcing klog dependency on other projects.

I think there is definitely also need to simplify the current logging interface and make it more plug able. This is inline with current direction of klog (introduction of SetLogger) and structured logging migration. Exact plan for replacement should be definitely discussed within a a KEP, but I don't think klog should remain as an the client interface used in Kubernetes (direct calls to klog.InfoS). We may keep the log format to remain backward compatible, but I expect that when we start introducing more advanced features (e.g. https://github.com/kubernetes/kubernetes/issues/91633) it will become not worth maintaining.

**Comment by serathius** [comment]
ping @liggitt

**Comment by mrbobbytables** [comment]
/assign
(to get it on my radar again)

File: wg-structured-logging/charter.md
Comment (mrbobbytables) [reply wg-structured-logging/charter.md:50]: > klog is not very favored dependency In golang community.

I would agree there, I do have a couple follow up thoughts:
- Including the klog replacement does significantly increase the stakeholders involved. They don't necessarily have to be called out as stakeholder sigs, but it's used by pretty much everything both in k/k and the majority of subprojects - The KEP will likely need a lot time to get community input
- Would this WG also be helping shepherd the migration to that klog replacement? The scope gets a bit wishy washy there with it being included.

Comment (serathius) [reply wg-structured-logging/charter.md:50]: I can agree that putting this in the charter unnecessarily complicates the scope.

**Comment by serathius** [comment]
PTAL @liggitt @mrbobbytables

**Comment by mrbobbytables** [comment]
Thanks!
/lgtm
/hold
for @liggitt and any other steering members to weigh in :)

**Comment by dims** [comment]
LGTM, looks like one pending question from @liggitt would be good to wrap up.

File: wg-structured-logging/charter.md
Comment (serathius) [reply wg-structured-logging/charter.md:28]: ping @liggitt

Comment (liggitt) [reply wg-structured-logging/charter.md:28]: > To make sure I understood you correctly, when talking about "upstream" dependencies do you mean:
> 
> * Kubernetes shared libraries in ./staging/src/k8s.io directory? or
> * Third party libraries outside of k/k repo?

I mean both:
1. things inside the staging/src/k8s.io
2. things in k8s.io/kubernetes/vendor that log or output to os.Stdout / os.Stderr

It seems like all have to be addressed to make the core components output completely structured... number 2 is a significantly bigger task/scope

Comment (serathius) [reply wg-structured-logging/charter.md:28]: I don't think that number 2 is as big as it might seem at first sight. Issues with third party libraries were already present before structured logging effort. klog also standardizes a log output format by using a standard header. For example we are injecting a klog based implementation of zap logger so Etcd client will write logs in klog format.

We already don't provide any guarantee that all logs have klog format, so the question should be "are we ok with the current state of log format completeness or should we do any steps to improve it?"

As charter specifies wg should allow "users to efficiently consume (logs)". WG should measure impact of third party logging on quality of log output and improve it if needed, but I don't it will be easy or profitable to target 100% coverage.

File: wg-structured-logging/charter.md
Comment (liggitt) [inline wg-structured-logging/charter.md:56]: sig or wg? do other wg's link to sig-governance or wg-governance?

File: wg-structured-logging/charter.md
Comment (liggitt) [reply wg-structured-logging/charter.md:28]: ok, maybe clarify this out-of-scope bit to "Logging outside of the kubernetes/kubernetes repository", since this WG does care about logging by things inside kubernetes/kubernetes vendor

File: wg-structured-logging/charter.md
Comment (serathius) [reply wg-structured-logging/charter.md]: Done. replaced with wg-governance and fixed the link

File: wg-structured-logging/charter.md
Comment (serathius) [reply wg-structured-logging/charter.md]: Updated, this will additionally add the other non-core components like kubeadm, but I think we should have uniform logging withing k/k.

Comment (liggitt) [reply wg-structured-logging/charter.md]: oh, that's a good point... I was focused on the impacts of logging from dependencies of the core components

the considerations for CLI components like kubeadm and kubectl are different... I see you described those as being out of scope in https://github.com/kubernetes/community/pull/5770/files#r644257373 ... would copying that line into the "out of scope" section and including kubeadm help clarify the WG focus?

> Binaries like kubectl and kubeadm will not be directly targeted nor we plan to change application logs that kubectl logs reads.

Comment (neolit123) [reply wg-structured-logging/charter.md]: we also saw a number of structured logging PRs for /test.

File: wg-structured-logging/charter.md
Comment (serathius) [inline wg-structured-logging/charter.md:27]:
```suggestion
- Logging outside the kubernetes/kubernetes repository
- Non core Kubernetes component binaries like kubectl and kubeadm
- Application logs read by kubectl
```

**Comment by serathius** [comment]
@liggitt PTAL

**Comment by liggitt** [comment]
this lgtm, thanks for the clarifications

not sure if you want to squash before merge

**Comment by serathius** [comment]
Squashed

**Comment by serathius** [comment]
ping @mrbobbytables

**Pull request merged**
